### DISCUSSION TITLE
hasMany not showing correctly on item page

### DISCUSSION THREAD

**Discussion by Glenf** [statamic-rad-pack/runway#264] [Help] [answered]
Hi,

I'm having issue where hasMany field doesn't show items correctly on parent items page, but it works fine on listing.

Works on listing looks like this:
![image](https://user-images.githubusercontent.com/1451152/236625777-551dd403-71a5-460f-a17c-700b0a074aa2.png)

And parent item page looks like this:
![image](https://user-images.githubusercontent.com/1451152/236625824-3011b107-0f19-480f-bc60-3f7646f0c99e.png)

It is the same issue with other field modes also.

It does get the correct amount of "has_many" items but I'm unable to get anything visible/selectable there. On listing page I can click individual items (dates) and links work. Is there a debugger or something I could turn on to see what the issue is?

**Comment by duncanmcclean** [comment]
If you switch to the mode to 'default' from 'table', does that change anything?

If so, are you able to share code?

**Reply by Glenf** [reply]
It doesn't. 
![image](https://user-images.githubusercontent.com/1451152/236626944-6f0a306e-ac6d-4b9c-80d1-b441021be789.png)

What part of code would you want to see?

**Reply by duncanmcclean** [reply]
1. Are you on the latest version of Runway?
2. Can you share your Runway config?
3. Can you share the blueprint for both resources you're looking at? (the one the HasMany field is on and the one you're seeing via the HasMany field)
4. What does the Eloquent relationship look like for that relationship? 
5. How did you add the relations there - using your own code added via the blueprint?

**Reply by Glenf** [reply]
Runway version is ~~4.2.5~~ updated to 4.2.8

I've redacted some most parts of blueprints, as those are working correctly in listings. I assume the connections are important parts here. I also have blueprints set up in config mainly for debugging reasons, they live in database and changes to that structure is a bit cumbersome. 

```php
'resources' => [
	\App\Models\Elokuva::class => [
		'name' => 'Elokuva',
		'handle' => 'elokuva',
		'title_field' => 'NIMI',
		'order_by_direction' => 'desc',
		'cp_icon' => 'video',
		'blueprint' => [
			"sections" => [
				"main" => [
					"display" => "Main",
					"fields" => [
						[
							"handle" => "NIMI",
							"field" => [
								"type" => "text",
								"display" => "Nimi",
								"instructions_position" => "above",
								"visibility" => "visible",
								"input_type" => "text",
								"antlers" => false
							]
						],
						[
							"handle" => "naytos",
							"field" => [
								"mode" => "default",
								"resource" => "naytos",
								"create" => true,
								"display" => "Näytökset",
								"type" => "has_many",
								"icon" => "has_many",
								"instructions_position" => "above",
								"visibility" => "visible",
							]
						]
					],
					"__count" => 0
				]
			]
		]
	],
	\App\Models\Naytos::class => [
	'name' => 'Näytökset',
	'handle' => 'naytos',
	'order_by_direction' => 'desc',
	'blueprint' => [
		"sections" => [
			"main" => [
				"display" => "Main",
				"fields" => [
					[
						"handle" => "SNUMERO_ID",
						"field" => [
							"max_items" => 1,
							"mode" => "typeahead",
							"create" => false,
							"display" => "Elokuva",
							"type" => "belongs_to",
							"icon" => "belongs_to",
							"listable" => "hidden",
							"instructions_position" => "above",
							"visibility" => "visible",
							"resource" => "elokuva"
						]
					],
					[
						"handle" => "ALKAMISAIKA",
						"field" => [
							"type" => "date",
							"display" => "Alkamisaika",
							"time_enabled" => true,
							"time_seconds_enabled" => false,
							"width" => 50
						]
					],
					[
						"handle" => "LOPPUMISAIKA",
						"field" => [
							"type" => "date",
							"display" => "Loppumisaika",
							"time_enabled" => true,
							"time_seconds_enabled" => false,
							"width" => 50
						]
					],
				],
				"__count" => 0
			]
		]
	]
]
```

Eloquent relationship in models are as following:

Naytos.php
```php
    public function elokuva(): BelongsTo
    {
        return $this->belongsTo(\App\Models\Elokuva::class, 'SNUMERO_ID', 'SNUMERO_ID');
    }
```

Elokuva.php
```php
protected $primaryKey = 'SNUMERO_ID';

[...]

public function naytos(): HasMany
{
    return $this->hasMany(\App\Models\Naytos::class, 'SNUMERO_ID', 'SNUMERO_ID');
}
```

Relations were made by me (a long time ago). This code base has been evolving about 15 years, and it shows in database. The data Runway is using has capitalized column naming in database. Not sure if that matters here.

**Reply by duncanmcclean** [reply]
Thanks! I've setup a fresh site, trying to replicate your setup as best I can but everything's working on my end....

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/19637309/236647278-1ed918c3-f6fb-4909-9fc3-ff1e149eaf5c.png">

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/19637309/236647286-3034566f-7f4e-4970-ab1c-e444503ebe12.png">

*^ works in both table & default mode*

I've pushed the code for my fresh site to this repo if you want to pull it down and take a look: https://github.com/duncanmcclean/runway-264

**Reply by Glenf** [reply]
Thanks for trying to repro this. I'm convinced that Runway works as it should and the issue lies somewhere in my code. I just have to narrow it down, so I can fix it :).

**Reply by Glenf** [reply]
Ok, I managed to narrow the issue down. Changing the primary key to something else than default brings the issue up. 

\App\Models\Naytos.php
```
$primaryKey = 'NAYTOS_ID';
```
this makes the issue appear.

**Comment by duncanmcclean** [comment, marked as answer]
It looks like Runway had the `id` column hard-coded in the HasMany fieldtype when it looks up the models before returning it to the publish form.

Fixed in [v4.2.9](https://github.com/duncanmcclean/runway/releases/tag/v4.2.9) and [v5.0.3](https://github.com/duncanmcclean/runway/releases/tag/v5.0.3).
